### ISSUE TITLE
Firebase auth leaks promises

### ISSUE THREAD

**Issue by gautambt** [firebase/firebase-js-sdk#1095]
### [REQUIRED] Describe your environment

  * Operating System version: Ubuntu 16.04 
  * Browser version: Chromium 67.0.3396.99 Built on Ubuntu , running on Ubuntu 16.04
  * Firebase SDK version:5.3.1
  * Firebase Product: Auth (auth, database, storage, etc)

### [REQUIRED] Describe the problem

Firebase auth seems to have a memory leak even when using the indexdedb storage engine for storing auth token.

#### Steps to reproduce:
1) Open the link https://firebase-issue-sandbox-emyrtq.stackblitz.io in chrome
2) Open devtools and run profiler. You can see that memory is allocated every 800ms and is retained forever (Clicking the collect garbage button does not free it). I've attached a screenshot; 
![firebase-auth-memory-leak](https://user-images.githubusercontent.com/203972/43769771-2f1662f8-9a58-11e8-92db-87567da0086b.png)

I've spent some time trying to debug the issue. The issue goes away if indexeddb is disabled. Also the 800ms correlates with fireauth.storage.IndexedDB.POLLING_DELAY_ in packages/auth/src/storage/indexeddb.js

The repeat function in startListeners_ seems to be the problem. I suspect that an infinite chain of promises is being created in repeat as repeat calls itself recursively. This patch fixes the issue for me:

```
diff --git a/packages/auth/src/storage/indexeddb.js b/packages/auth/src/storage/indexeddb.js
index dcd71ec..8d7b744 100644
--- a/packages/auth/src/storage/indexeddb.js
+++ b/packages/auth/src/storage/indexeddb.js
@@ -603,7 +603,7 @@ fireauth.storage.IndexedDB.prototype.startListeners_ = function() {
                 });
           }


         })
-        .then(repeat)
+        .then(() => setTimeout(repeat))
         .thenCatch(function(error) {
           // Do not repeat if cancelled externally.
           if (error.message != fireauth.storage.IndexedDB.STOP_ERROR_) {
```
If the analysis seems correct and it is okay to fix the problem this way, I will go ahead and raise a pull request.

**Comment by gautambt** [comment]
Any thoughts/comments?

**Comment by bojeil-google** [comment]
I took another closer look.
So the continuous polling is required to catch external changes to indexedDB. This includes a user signing in or out from another tab or window. This is unavoidable and is not considered a memory leak. Overall the memory usage should be stable. Each time a cycle is completed, the intermediate variables should eventually be garbage collected. However, the next cycle will repeat the same process and the same amount of memory will be reallocated. The important thing is that we keep memory usage stable.

**Comment by gautambt** [comment]
@bojeil-google: I left the tab overnight and memory seems to continuously increase. Hitting the collect garbage button does nothing. I suspect something similar to what is described in https://alexn.org/blog/2017/10/11/javascript-promise-leaks-memory.html is happening with goog.Promise.

**Comment by bojeil-google** [comment]
I have a sample rewrite of the logic to not use infinite promise chaining: https://github.com/firebase/firebase-js-sdk/pull/1121
Can you check if this makes a difference for you?

**Comment by avickers** [comment]
This seems to be the root cause of #1035, so if the patch works they can both be closed.

**Comment by gautambt** [comment]
Will be checking the patch today.

**Comment by samtstern** [comment]
@gautambt says that this is fixed by #1121, so closing the issue.  Thanks all!
